### ISSUE TITLE
NAT Gateway configuration

### ISSUE THREAD

**Issue by NShaforostov** [epam/cloud-pipeline#2232]
## Background

Currently, if the Cloud Pipeline Platform is deployed in some private subnet, it could be quite difficult for the admin to expose a network endpoint for some service. This requires manual execution of a number of tasks on the Platform Core instance and, accordingly, may lead to errors.
It would be convenient to allow the management (creating/removing) network routes from the GUI - this would save from wasting time and allow to avoid possible errors in configurations.

### GUI

Add a new form to the **System Settings** that will allow admins to create/remove network routes from private subnet(s) to the external network resources.
By default, this form shall display:
- the list of existing routes (port forwarding map). Each route record should contain:
  - info about external resources:
    - server name
    - IP
    - port
  - info about corresponding internal config (mapping details):
    - service name
    - service IP
    - port
  - button to remove the route
- button to add a new route
- disabled button to save changes

Approximate view of that form:
![image](https://user-images.githubusercontent.com/45459424/138140678-cc9dae56-b305-40d8-bc41-f279cadf9ca0.png)

When admin clicks the **Add route** button - new field shall appear:
- text field to specify a server name of the adding service
- control to specify an IP of the adding service:
  - text field to specify IP manually
  - button to autoresolve IP by specified server name. This button should be disabled until the server name is not specified.
  Admin can specify IP manually or use this button - for the last case, we should implement the mechanism to resolve an IP and output it to the IP field
- control to specify a port of the adding service:
  - text field to specify a port
  - button to add another port. Admin shall have the ability to add any count of ports for one server name.
  By click this button - additional text field should appear under the first port field and a control to remove that additional port
- control to remove these fields

Approximate view of that form:
![image](https://user-images.githubusercontent.com/45459424/138140335-74a3ddac-0780-4af0-8218-da96550f4ec5.png)

Admin should specify all fields. The **Save** button should become enabled, e.g.:
![image](https://user-images.githubusercontent.com/45459424/138140138-9044026c-344a-4278-a55e-c22ce98f2acb.png)
Admin should click it to save changes. After the saving, the routing list shall be refreshed. If the form will be closed without click that button - all changes should be canceled (the corresponding warning message should be shown).

In case, when the route record was added - the new `Kubernetes` service, that handles requests from the private subnet and passes it to the "Port forwarding service" shall be created, the corresponding records shall be added to mapping configs. See details [below](#route-creation).

In case, when the route record was removed - the corresponding records shall be removed from mapping configs and the corresponding `Kubernetes` service, that handles requests from the private subnet and passes it to the "Port forwarding service", shall be deleted. See details [below](#route-remove).

Additionally:
- admin shall have the ability to add several routes at once
- admin shall have the ability to add several routes for a single server name but for different ports:
  - by adding several ports to a single route record
  - by adding several route records with the same server name but different ports
  - by adding route record(s) with the server name, that already exists in the routes list, but with another ports (differ from existing)
  In all these cases, for each server name only one `Kubernetes` service shall be used (no need to create a separate services for different ports).
- if any route was removed from the list - the **Save** button should become enabled. Such changes shall be saved as well
- there shall not be the ability to **Save** changes if there is any route with an empty (non-filled) field

> As for the placement of the described form:
> I suggest to rename the existing settings tab **SYSTEM LOGS** to **SYSTEM MANAGEMENT** and create two subtabs in it - for  Logs (for previous **SYSTEM LOGS** content) and NAT Gateway for described form (the rendering should be the same as for **PREFERENCES** subtabs e.g.).
> ![image](https://user-images.githubusercontent.com/45459424/138135271-8ee3ded8-4609-44c8-81c5-d75f463c2a8e.png)

<h3 id='route-creation'>Route creation flow</h3>

Here:
- `<SERVER_NAME>` - external server name (_getting from the GUI_)
- `<SERVER_IP>` - external server IP (_getting from the GUI_)
- `<PORT>` - external resources port (_getting from the GUI_)
- `<TARGET_IP>` - internal service IP 
- `<TARGET_PORT>` - port forwarding service port
- `<SPEC_PATH>` - path to the spec file

Adding a new route should include the following steps at the Platform Core instance:

1. Creating a new `Kubernetes` service, that handles requests from the private subnet and passes it to the "Port forwarding service":
  - create a spec file with the content like:
  ```
  apiVersion: v1
  kind: Service
  metadata:
    labels:
      cloud-pipeline/cp-tinyproxy: "true"
    name: cp-tinyproxy-<SERVER_NAME>
    namespace: default
  spec:
    ports:
      protocol: TCP
      port: <PORT>
      targetPort: <TARGET_PORT>
      name: cp-tinyproxy-<SERVER_NAME>-<PORT>
    selector:
      cloud-pipeline/cp-tinyproxy: "true"
  ```
  **_Note_**: `<TARGET_PORT>` shall be generated as unique unoccupied port number that is greater than `1000`
  - create a service from the spec:
  `kubectl create -f <SPEC_PATH>`
  - check that service is created and get `<TARGET_IP>` (value from the CLUSTER-IP column in the command result):
  `kubectl get svc cp-tinyproxy-<SERVER_NAME>`
2. Creating a new DNS entry, that points to the created `kubernetes` service:
  For that, open the internal DNS configuration map (`cp-dnsmasq-hosts`) in `kube-system` namespace and add the following line to the configuration map "hosts" section: `<TARGET_IP> <SERVER_NAME>`
3. Creating a "Port forwarding service" rule, which maps the traffic from the `kubernetes` service to the external service.
  For that, open the Platform global configuration map (`cp-config-global`) and add a port forwarding rule to the `CP_TP_TCP_DEST` entry (separated by the comma from the previous values).
  Format of the entry should be: `<TARGET_PORT>:<SERVER_IP>:<PORT>`
4. Restarting the forward proxy to pickup the fresh configuration

<h3 id='route-remove'>Route removing flow</h3>

Removing the existing route should include the following steps at the Platform Core instance:
1. Removing a "Port forwarding service" rule, which maps the traffic from the `kubernetes` service to the external service.
2. Removing a DNS entry, that points to the `kubernetes` service.
3. Deleting the corresponding `kubernetes` service.
4. Restarting the forward proxy to pickup the fresh configuration

**Comment by rodichenko** [comment]
@NShaforostov @sidoruka "NAT Gateway" settings implemented (f8a4ae008cdcd4445e4b1932d3bd01c328f30f6d)

**Comment by sidoruka** [comment]
@mzueva @rodichenko can we please also add a "comment" field to each entry?
I.e. allow to store any random string in the kube-service annotation and pass it to the client.

**Comment by mzueva** [comment]
@sidoruka @rodichenko server backported to `release/0.16`

**Comment by rodichenko** [comment]
@sidoruka GUI backported to `release/0.16` (5a41cb429baead83269589bd6f531d6e98a0d608)

**Comment by maryvictol** [comment]
@Wedds 
*Comment* added at creation route disappears when route's *CREATION_SCHEDULED* status changes to *ACTIVE*.

**Comment by maryvictol** [comment]
@Wedds 
If several routes for a single server name but for different ports are added and user try to delete one of them, then this route remains in  list with status UNKNOWN until all routes with the same server name are deleted.

**Comment by Wedds** [comment]
If some routes are configured already in a certain deployment, corresponding services could be edited to go under the NAT gateway service control.

**1. Edit service annotations**
Information regarding route external IP/name and ports' statuses is stored there.

Common route annotations
- nat-route-ip: **IP**
- nat-route-name: **NAME**

Specific port annotations (each port of a route should have its own set of anotations):
- nat-route-current-status-<PORT_NUMBER>: ACTIVE
- nat-route-target-status-<PORT_NUMBER>: ACTIVE
- nat-route-last-update-<PORT_NUMBER>: 2021-12-03_15.00.00.000000

**2. Edit service label**
- `cloud-pipeline/cp-tinyproxy=true` - current value
- `cloud-pipeline/cp-tinyproxy-nat=true` - target value 

Label is used to pull services from k8s in `cp-api`
Key suffix is configurable via application.properties using value:
> nat.gateway.cp.service.name=${CP_API_NAT_PROXY_SERVICE_NAME:cp-tinyproxy-nat}

**Comment by maryvictol** [comment]
@rodichenko 
Please, add warning "You have unsaved changes. Continue?" at case of the user leaves the NAT Gateway page without saving changes.

**Comment by sidoruka** [comment]
@Wedds, it throws the following exception if there are services with the correct labels, but no annotations:
```
java.lang.NullPointerException: null
        at com.epam.pipeline.manager.cluster.NatGatewayManager.extractStringFromAnnotations(NatGatewayManager.java:896) ~[classes!/:0.16.0.11545.9b14c326593c30540b72237e52e206d2
1aee94b5]
        at com.epam.pipeline.manager.cluster.NatGatewayManager.buildRouteFromPort(NatGatewayManager.java:636) ~[classes!/:0.16.0.11545.9b14c326593c30540b72237e52e206d21aee94b5]
        at com.epam.pipeline.manager.cluster.NatGatewayManager.lambda$extractNatRoutesFromKubeService$18(NatGatewayManager.java:626) ~[classes!/:0.16.0.11545.9b14c326593c30540b7
2237e52e206d21aee94b5]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[?:1.8.0_302]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384) ~[?:1.8.0_302]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:1.8.0_302]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:1.8.0_302]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_302]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_302]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566) ~[?:1.8.0_302]
```

**Comment by NShaforostov** [comment]
**Docs** were added via #2453 and located [here](https://github.com/epam/cloud-pipeline/blob/develop/docs/md/manual/12_Manage_Settings/12.14._NAT_gateway.md).

**Comment by NShaforostov** [comment]
**Test cases** were created by #2443 and located [here](https://github.com/epam/cloud-pipeline/tree/develop/docs/testcases/UI/System_Management/2232-2444_NAT_Gateway).

**Comment by NShaforostov** [comment]
**e2e-tests** were implemented via #2449.

**Comment by NShaforostov** [comment]
**Daily e2e-tests results** are located and updated [here](https://github.com/epam/cloud-pipeline/blob/e2e-reports/e2e/reports/AWS/test/classes/com.epam.pipeline.autotests.NATGatewayTest.html) (for `AWS`).
